### PR TITLE
Add http client router feature

### LINKED ISSUE #12: HttpClientRouter
A client accepting a map of regex -> httpclient. Given a specific request, it will check the uri of it and redirect to a specific http client implementation.

This allow to have a single pipeline for multiple library, but still allows using different client for some requests.

### PR DESCRIPTION
**Pull request by joelwurtz**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #12
| Documentation   | php-http/documentation#103
| License         | MIT

#### What's in this PR?

Explained in #12

#### Checklist

- [x] Updated CHANGELOG.md to describe BC breaks / deprecations | new feature | bugfix
- [x] Documentation pull request created (if not simply a bugfix)
- [x] #11 Need to be merged before this
- [x] Matcher should be added in php-http/message (and have some basics implementations)
- [x] Squash before merge

### CODE REVIEW

File: src/RequestMatcher.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Http\Client\Common;
+use Psr\Http\Message\RequestInterface;
+
+/**
+ * RequestMatcher allow to tell if a PSR7 Request matches a specific strategy
+ *
+ * @author Joel Wurtz <joel.wurtz@gmail.com>
+ */
+interface RequestMatcher
```
Comment (dbu) [inline src/RequestMatcher.php]: did we not already add a RequestMatcher somewhere?

File: src/HttpClientFlexible.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+use Http\Client\HttpAsyncClient;
+use Http\Client\HttpClient;
+
+/**
+ * A flexible http client, which implements both interface and will emulate
+ * one contract, the other, or none at all depending on the injected client contract.
+ *
+ * @author Joel Wurtz <joel.wurtz@gmail.com>
+ */
+class HttpClientFlexible implements HttpClient, HttpAsyncClient
```
Comment (dbu) [inline src/HttpClientFlexible.php]: i think this name is a bit weird. maybe HttpSyncAsyncClient ? or any better ideas?

File: src/RequestMatcher.php (hunk 1 above)
Comment (sagikazarmark) [inline src/RequestMatcher.php]: We did, message. See https://github.com/php-http/client-common/pull/14#issuecomment-203115143

File: src/HttpClientFlexible.php (hunk 2 above)
Comment (sagikazarmark) [inline src/HttpClientFlexible.php]: This is already in the package with FlexibleHttpClient. This PR needs rebase.

File: src/HttpClientRouter.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    /**
+     * Add a client to the router.
+     *
+     * @param HttpClient|HttpAsyncClient $client
+     * @param RequestMatcher             $requestMatcher
+     */
+    public function addClient($client, RequestMatcher $requestMatcher)
+    {
+        $this->clients[] = [
```
Comment (sagikazarmark) [inline src/HttpClientRouter.php]: Do you think we could find a better way to store these. Although it's an internal detail, it feels a little but weird.

**Comment by sagikazarmark** [comment]
Did some improvments. Anyone review?

File: src/HttpClientRouter.php (hunk 3 above)
Comment (dbu) [inline src/HttpClientRouter.php]: having a member client and a member matcher would make more sense.

re-reading this i am confused about this code. we have a bunch of matcher, but only one single client? if none of the matchers match, we throw an exception, if any matches we always use the same client. that is not what i would expect from a router. i would have expected this to be an array of matcher, client and if a matcher matches, that client is used...

Comment (dbu) [inline src/HttpClientRouter.php]: the code also indicates that this was originally the idea.

Comment (dbu) [inline src/HttpClientRouter.php]: oh sorry. i misread the code. it does what i would expect. i don't see how we could store these differently. its a list of pairs [matcher, client] - i would not know how to store that differently. you could try some spl hashmap that can use objects as key if that exists, but i think its quite straightforward.

**Comment by dbu** [comment]
looks good to me. can we squash the commits a bit?

also, according to the description we have no documentation yet and we should update the changelog.

File: src/HttpClientRouter.php (hunk 3 above)
Comment (sagikazarmark) [inline src/HttpClientRouter.php]: > i would have expected this to be an array of matcher, client and if a matcher matches, that client is used...

I think this is what's happening right now. Isn't it?

That one client is only used as a decorator in order to correctly emulate a client if necessary.

**Comment by sagikazarmark** [comment]
Will do.

**Comment by sagikazarmark** [comment]
Changelog updated, doc PR ready.

**Pull request merged**